install_requires list

weaverba137 opened the pull request:
This PR

 * Closes #225 
 * Adds a specific, explicit list notation for specifying multiple install dependencies.

astrofrog commented on setup.py:
```diff
@@ -125,7 +125,7 @@
       description=DESCRIPTION,
       scripts=scripts,
-      install_requires=metadata.get('install_requires', 'astropy').strip().split(),
+      install_requires=eval(metadata.get('install_requires', "['astropy']")),
```
I strongly feel that we shouldn't use eval. If this is the only way to parse it, I would vote for using the simple comma-separated format rather than a Python list.

pllim replied:
:+1: to **not** use `eval()`.

astrofrog commented:
I object to the Python list format and would much prefer the simple comma-separated format, both for simplicity of parsing and because of backward compatibility.

weaverba137 commented:
Noted!  What is the recommended way to handle comma-separation though, because:
```python
'astropy,scipy,matplotlib'.split(',')
```
produces a different list from
```python
'astropy, scipy, matplotlib'.split(',')
```
In other words, how much "slop" do we allow with the commas?

astrofrog commented:
What about simply doing `` [s.strip() for s in install_requires.split(',')]``?

weaverba137 commented:
Switched to comma-separated.

bsipocz commented on setup.py:
```diff
@@ -125,7 +125,7 @@
       description=DESCRIPTION,
       scripts=scripts,
-      install_requires=[s.strip() for s in metadata.get('install_requires', 'astropy').split(',')],
+      install_requires=[s.strip() for s in metadata.get('install_requires', 'astropy').strip().split(',')],
```
if this second strip() necessary?

weaverba137 replied:
You tell me. That same `.strip()` was in the original setup.py file, so I'm just replicating it.

weaverba137 commented:
You tell me.  That same `.strip()` was in the original setup.py file, so I'm just replicating it.

bsipocz commented:
You've have that ``.strip()`` already at the beginning of the list comprehension.

weaverba137 commented:
Please go back and look at the original setup.py file.

bsipocz commented:
@weaverba137 - You have double ``strip()`` now, one at the beginning and one at the end of the list comprehension. One of those does nothing I think.

Originally the was only one as there was no list comprehension. At least this is what I see on GH when comparing to the original:

```
-      install_requires=metadata.get('install_requires', 'astropy').strip().split(),
+      install_requires=[s.strip() for s in metadata.get('install_requires', 'astropy').strip().split(',')],
```

bsipocz commented:
(I can remove the last commit manually before merging, just wanted to know whether there was a reason for it.)

astrofrog commented:
Note that we should put merging this on hold as the cookiecutter branch is now set up and we should no longer commit directly to master. Instead we should make this change in the cookiecutter branch. We still need to set up docs about how to use the cookiecutter branch, but once that's done it should be clear how to do this.

astrofrog commented:
🙏 let's keep it civil :)

I agree with @bsipocz that the original strip can now be removed since it is redundant.

weaverba137 commented:
Sure, I agree that it *can* be removed, I just wanted to emphasize that it's in there to try to make this a minimal change.

weaverba137 commented:
To move things along, I've removed the extra `.strip()` method.

weaverba137 commented:
Ping!

weaverba137 commented:
Does this need to be merged onto some other branch?

weaverba137 commented:
Can someone please let me know how to proceed with this?

bsipocz commented:
I've changed the base to the cookiecutter branch, but this still needs some cleanup as the changes there should point to the files under the ``{{ cookiecutter.package_name }}`` directory.

weaverba137 commented:
OK, I'm working on that.  I noticed that `{{ cookiecutter.package_name }}/setup.cfg` no longer contains `install_requires` at all.   But it looks like adding 
```
install_requires = {{ cookiecutter.install_requires }}
```
would be sufficient.  If this item is added, will it actually be configured automagically, or does that require an additional change to some other configuration file?

weaverba137 commented:
I updated cookiecutter.json. The only remaining question would be how does someone creating a new package know that `install_requires` takes a comma-separated list?

weaverba137 commented:
Ping.

> The only remaining question would be how does someone creating a new package know that install_requires takes a comma-separated list?

bsipocz commented:
@Cadair planned to add docs for the template usage with cookiecutter, ultimately this info should go in there I suppose. I'm not sure that leaving comments in the json file would make any sense.

weaverba137 commented:
@Cadair, could you please advise on how to proceed with this?  The only thing this PR needs is where to put the documentation?

weaverba137 commented:
@Cadair, could you please advise on how to proceed with this? The only thing this PR needs is where to put the documentation?

I don't want this PR closed by a bot due to lack of response.

bsipocz commented:
Thank you @weaverba137 for your patience. I'm merging this now while the documentation is figured out.

weaverba137 commented:
Cheers, thank you!
